more detailed docstring for read! and write

mgkuhn opened the pull request:
The old `read!` docstring failed to describe how many bytes the
function reads and what its return value is.

Seelengrab commented on base/io.jl:
```diff
@@ -475,10 +475,32 @@ read(filename::AbstractString, args...) = open(io->read(io, args...), convert(St
+    read!(stream::IO, a::Ref)
+    read!(filename::AbstractString, a::Union{AbstractArray, Ref})
+
+Read `sizeof(a)` bytes of binary data from an I/O stream or file,
```
Does the function even work for non-isbits?

mgkuhn replied:
Yes, we will probably have to distinguish in the docstring both cases of `isbitstype(T)`. (This should probably also be done for a docstring covering `write(a::AbstractArray{T,N})`)

If `isbitstype(T)==false`, then `read!(..., a:AbstractArray{T,N})` just runs
```
for i in eachindex(a)
    a[i] = read(s, T)
end
```
i.e. falls back to whatever `read()` method was provided for type `T`, and the number of bytes read will depend on what that does. Also, it seems `read!(..., a:Ref{T})` currently doesn't have such a fallback and therefore is only intended for `isbitstype(T)==true`.

Seelengrab replied:
Right, that's what I mean - non-bitstypes can't really be read like that, because of the pointers contained within. They need to be properly deserialized and recreated in a more complicated manner.

mgkuhn committed:
more detailed docstrings for read! and write

The old `read!` docstring failed to describe how many bytes the
function reads and what its return value is.

The old `write` docstring didn't cover arrays.

I also documented `read!(..., Ref{T})`: I first wasn't sure if this
was meant to be documented, but then saw that it was already mentioned
in the docstring of `write`.

mgkuhn commented:
I've now updated this docstring PR:
* I've used for `read!` phrasing similar to what @stevengj suggested
* I've added a note that `read!(..., Ref{T})` is only intended for primitive types. I first wasn't sure whether it should be documented at all (as it seems an unsafe function when used with non-primitive types), but then saw that it was already mentioned in the docstring of `write`.
* The `write` docstring now also covers writing arrays.

stevengj commented on base/io.jl:
```diff
@@ -475,10 +484,39 @@ read(filename::AbstractString, args...) = open(io->read(io, args...), convert(St
+`a::AbstractArray{T}` is read with a call to `read(stream, T)`.
+
+The `Ref{T}` variant is not intended for use if `isbitstype(T)` is
+false.
```
This is seems broken to me.  It really seems like we should have a safe fallback:
```jl
function read!(s::IO, x::Ref{T}) where {T}
    if isbitstype(T)
        unsafe_read(s, x, Core.sizeof(T))
    else
        x[] = read(s, T)
    end
    return x
end
```

mgkuhn replied:
@stevengj I think I agree. Since a `Ref{T}` is essentially a 0-dimensional array of length 1, `read!` should work the same for `Ref` and `Array`. That then will also simplify its docstring.

vtjnash replied:
Thanks for pointing that out. I made a PR to fix that https://github.com/JuliaLang/julia/pull/51949

vtjnash commented on base/io.jl:
````diff
@@ -273,12 +276,18 @@ julia> write(io, "Sometimes those members") + write(io, " write documentation.")
-User-defined plain-data types without `write` methods can be written when wrapped in a `Ref`:
+Arrays:
+```jldoctest
+julia> write(stdout, [0x41 0x42; 0x43 0x44], 0x0a)
````
```suggestion
julia> write(stdout, UInt8['A' 'B'; 'C' 'D'], '\n')
```

vtjnash commented on base/io.jl:
```diff
@@ -475,10 +484,39 @@ read(filename::AbstractString, args...) = open(io->read(io, args...), convert(St
+Read `length(a)` consecutive values of type `T` from an I/O stream or
+file into `a`, returning `a`.
+
+If `a` is an `Array` and `isbitstype(T)` is true, this function reads
+`size(a)` bytes directly into memory. Otherwise, each element of
+`a::AbstractArray{T}` is read with a call to `read(stream, T)`.
```
```suggestion
If `isbitstype(T)` is true, this function reads `sizeof(a)` bytes
directly into memory. Otherwise, each element of `a` is populated
sequentially by a call to `read(stream, T)`.
```

N5N3 replied:
Not sure if `sizeof(a)` is accurate here.
`Base.sizeof` is not defined for most `AbstractArray` and it falls back to `Core.sizeof` by default.

BTW, if we expect `AbstractArray`s to behave like `Array`s, then the optimization for `StridedArray`s needs to check `elsize(A) == elsize(Array{eltype(A)})` to avoid possible padding mismatch.

vtjnash replied:
yeah, maybe `sizeof(T) * length(a)` bytes would be better then? (more accurate yet would be `Base.aligned_sizeof`, but that seems a bit overly pedantic to me)

DilumAluthge commented:
This PR has had no activity for two years+.

@mgkuhn Are you still interested in working on this PR?
